src/ArraySlice.hpp

```diff
@@ -280,6 +280,82 @@ class ArraySlice
 
 };
 
+/**
+ * @tparam T The type of @p value.
+ * @tparam LAMBDA the type of the function @p f to apply.
+ * @brief Apply the given function to the given value.
+ * @param value the value to pass to the function.
+ * @param f the function to apply to the value.
+ */
+DISABLE_HD_WARNING
+template< typename T, typename LAMBDA >
+LVARRAY_HOST_DEVICE
+void forValuesInSlice( T & value, LAMBDA && f )
+{ f( value ); }
+
+/**
+ * @tparam T The type of values stored in @p slice.
+ * @tparam NDIM the dimension of @p slice.
+ * @tparam USD the unit stride dimension of @p slice.
+ * @tparam INDEX_TYPE the integer used to index into @p slice.
+ * @tparam LAMBDA the type of the function @p f to apply.
+ * @brief Iterate over the values in the slice in lexicographic order.
+ * @param slice the slice to iterate over.
+ * @param f the function to apply to each value.
+ */
+DISABLE_HD_WARNING
+template< typename T, int NDIM, int USD, typename INDEX_TYPE, typename LAMBDA >
+LVARRAY_HOST_DEVICE
+void forValuesInSlice( ArraySlice< T, NDIM, USD, INDEX_TYPE > const & slice, LAMBDA && f )
+{
+  for( INDEX_TYPE i = 0 ; i < slice.size( 0 ) ; ++i )
+  {
+    forValuesInSlice( slice[ i ], f );
+  }
+}
+
+/**
+ * @tparam T The type of values stored in @p slice.
+ * @tparam INDICES variadic pack of indices.
+ * @tparam LAMBDA the type of the function @p f to apply.
+ * @brief Apply the funtion @p f to the value @p value also
+ *        passing @p f any indices used to reach @p value.
+ * @param slice the slice to iterate over.
+ * @param f the lambda to apply to each value.
+ * @param indices the previous sliced off indices.
+ */
+DISABLE_HD_WARNING
+template< typename T, typename LAMBDA, typename ... INDICES >
+LVARRAY_HOST_DEVICE
+void forValuesInSliceWithIndices( T & value, LAMBDA && f, INDICES const ... indices )
```

Comment:
I am a little confused by the indices parameter. In the calls to this function, it appears that this parameter is not given an argument?? The `indices` arguments that I see in the call are arguments to the lambda, not this function?

Reply:
Correct, this is the base case for the following function. This allows you to do something like
```c++
Array< int, 3 > x( 2, 3, 4 );
Array< int, 2 > y( 2, 3, 4 );
...
// Adds x_ijk to y_ijk
forValuesInSliceWithIndices( x, 
  [&]( int const & value, auto const... indices )
  {
    y( indices... ) += value;
  }
);
```
Which is pretty trivial, but when the loop is in a templated method taking that can take in an arbitrary dimension array with any permutation it is very useful.